### PR TITLE
8339512: [TestBug] Convert graphics tests to JUnit 5

### PR DESCRIPTION
**Pull request by lukostyra**
This PR converts all tests in `modules/javafx.graphics` to use JUnit5.

## Details

Trivial changes resolved by first four commits:
- Import changes to use `org.junit.jupiter` package instead of `org.junit` or `junit.framework` and all changes that follow it (ex. `@Before` -> `@BeforeEach`)
- Message Strings in `assert*` calls moved to the end to follow 5's argument order
- Some simple Parametrized tests (aka. those which only did a straightforward assignment of test parameters) adjusted to use JUnit5's `@ParametrizedTest` and `@MethodSource` annotations
- Some test formatting unification - keeping annotations on a separate line from method definition
- Some tests used `@Rule` annotation which is missing from JUnit5. Luckily, all those uses just set `ExpectedException thrown = ExpectedException.none()` and then mid-test set it to some expected exception class. This was easily replaceable with `assertThrows`.

Non-trivial changes:
- Mutliple tests use common test bases and derive from it to check specific behaviors. Those include:
  - `DirtyRegionTestBase`
  - `CssMethodsTestBase`
  - `ObjectMethodsTestBase`
  - `OnInvalidateMethodsTestBase`
  - `PropertiesTestBase`
- `ServiceTestBase` & derivatives required some rework due to `ServiceExceptionTest` being a derived parametrized class while other `ServiceTestBase` were not parametrized
- `TransformOperationsTest` not only expanded 3 test parameters onto 6 different test class members, but also exposed its `getParams()` call to other testers - `AffineOperationsTest` and `TransformUtilsTest`
- Other classes used `@BeforeEach` annotation which accessed test parameters, those setup methods were called manually in-test as per discussion in #1561 

At the end I added a change converting `assertTimeout` uses I added to `@Timeout` annotation, as recommended in other PRs.

As for conversion practices and methods, most of these were discussed as part of #1561 so I will skip them here.

## Results

Below are results of the test conversion and runs of `gradle --info :graphics:test` on my machine:

| Category | JUnit4 ( 4647367ce ) | JUnit5 (this PR) |
| ------------ | ------ | ------ |
| Test count | 23272  | 23275 |
| Failures | 0 | 0 |
| Ignored | 99 | 102 |
| Successful % | 100% | 100% |

The difference in test count and ignored comes from `Popup_parentWindow_Test.java` where the entire test class was annotated as `@Ignored`. After conversion, the `@Disabled` annotation applies to each test individually rather than the whole class, so:
```
Popup_parentWindow_Test. initializationError
```
becomes
```
Popup_parentWindow_Test. testBasicAccess(Configuration)
Popup_parentWindow_Test. testBinding(Configuration)
Popup_parentWindow_Test. testGetBean(Configuration)
Popup_parentWindow_Test. testGetName(Configuration) 
```

Outside of that, the ignore list matches between both runs. I did not notice any regressions or other differences in test execution.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8339512](https://bugs.openjdk.org/browse/JDK-8339512): [TestBug] Convert graphics tests to JUnit 5 (**Enhancement** - P4)

### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Marius Hanl](https://openjdk.org/census#mhanl) (@Maran23 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1566/head:pull/1566` \
`$ git checkout pull/1566`

Update a local copy of the PR: \
`$ git checkout pull/1566` \
`$ git pull https://git.openjdk.org/jfx.git pull/1566/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1566`

View PR using the GUI difftool: \
`$ git pr show -t 1566`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1566.diff">https://git.openjdk.org/jfx/pull/1566.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1566#issuecomment-2358777495)

### CODE REVIEW

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/font/PrismFontFactoryTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/font/PrismFontFactoryTest.java:65]: this is good.

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionClipTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionClipTest.java:66]: I think junit5 also accepts a `List<Arguments>`

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java:333]: should probably avoid dropping creator/polluter from output

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java:431]: same comment here (and possibly elsewhere)
one thing you could do is to retain instance variables, just set them in the setup method - this should be safe, I don't think junit will invoke test methods for the same instance from multiple threads.

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/test/CssMethodsTestBase.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/test/CssMethodsTestBase.java:113]: I think this can be simplified: use Configuration directly, no need to wrap it into Arguments.
But this code will work too.

**Review by andy-goryachev-oracle — COMMENTED** [review]
first batch of comments.  
will continue tomorrow from ObjectMethodsTestBase.java

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java:76]: There will be plenty of uses of that all throughout this change, I wasn't 100% sure what JUnit5 expects or allows so I used the same `Stream<Arguments>` return type for parameter methods across all tests.

Comment (Maran23) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java:76]: This is fine and I think Stream<Argument> is what you will in most documentations as well, so good IMO.

File: modules/javafx.graphics/src/test/java/test/javafx/concurrent/ScheduledServiceTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/concurrent/ScheduledServiceTest.java:95]: could have inserted a newline after the annotation...

Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/test/java/test/javafx/concurrent/ScheduledServiceTest.java:95]: this particular test is not parameterized, so you could use `@BeforeEach` to setup the factory.  Same for the `ServiceLifecycleTest`.

it won't work for `ServiceExceptionTest` because it is parameterized.

but this code is ok too.

File: modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceLifecycleTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceLifecycleTest.java:2107]: minor: maybe `Class<? extends Throwable> exceptionClass` ?

File: modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceTestBase.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceTestBase.java:31]: minor: this line is probably unnecessary

File: modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceTestBase.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceTestBase.java:65]: I wonder if we should log a JBS ticket to fix parameterized tests by reverting `@BeforeEach`, and then simply refer to it?
This way, once junit supports that, we can easily grep the code.

(I've used a different comment for same purpose)

What do you think?

**Review by andy-goryachev-oracle — COMMENTED** [review]
next batch.  will continue with Node_effectiveOrientation_Test.java

**Comment by andy-goryachev-oracle** [comment]
on the other hand, @kevinrushforth mention that it is unlikely we would spend more time on parameterized tests, so please ignore my comment about JBS ticket.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/TreeShowingPropertyTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/TreeShowingPropertyTest.java:68]: the change may not be fully equivalent: the use of `Supplier` defers the moment of creation until each test is started.

it is very likely that the parameters() method is called once for each test, creating all the instances before any of the tests cases are executed.

this still might be ok because 
a) the tests are headless
b) the nodes are not inserted into the scene graph
c) instances are not reused anywhere

File: modules/javafx.graphics/src/test/java/test/javafx/scene/effect/BoxBlur_properties_Test.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/effect/BoxBlur_properties_Test.java:44]: you _could_ return the `List<Arguments>` directly

File: modules/javafx.graphics/src/test/java/test/javafx/scene/effect/EffectInputTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/effect/EffectInputTest.java:71]: (here and elsewhere, you could return the List directly)

File: modules/javafx.graphics/src/test/java/test/javafx/scene/layout/HBoxTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/layout/HBoxTest.java:47]: minor: newline?

File: modules/javafx.graphics/src/test/java/test/javafx/scene/layout/RegionCSSTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/layout/RegionCSSTest.java:685]: minor: indent

File: modules/javafx.graphics/src/test/java/test/javafx/scene/transform/AffineOperationsTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/transform/AffineOperationsTest.java:1929]: that's why it would be easier to return List<Affine> from `getParams`

File: modules/javafx.graphics/src/test/java/test/javafx/scene/transform/AffineOperationsTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/transform/AffineOperationsTest.java:2781]: this is one humongous test!!!

File: modules/javafx.graphics/src/test/java/test/javafx/scene/transform/TransformOperationsTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/transform/TransformOperationsTest.java:2944]: the trick with assigning instance fields might have helped here.
I discovered this too late myself, and now it's probably too late anyway...

File: modules/javafx.graphics/src/test/java/test/javafx/stage/WindowTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/stage/WindowTest.java:222]: this looks like unrelated change:
- why was the name changed?
- why change this method and not testSizeToSceneAfterShowing?

File: modules/javafx.graphics/src/test/jslc/com/sun/scenario/effect/compiler/lexer/TypeTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/jslc/com/sun/scenario/effect/compiler/lexer/TypeTest.java:35]: for other reviewers: I think this cleanup is good.

**Review by andy-goryachev-oracle — APPROVED** [review]
I looked at each line - this is one monumental piece of work, good job @lukostyra !

Areas that I checked explicitly:
- before -> beforeEach, etc.
- expected -> assertThrows, if same exception
- parameterized tests

I did not actually run the tests.

I left some minor comments, but overall this is good even as is.  Will re-approve if you decide to make fixes.

**Comment by andy-goryachev-oracle** [comment]
should we set a minimum of two reviewers?  please do not integrate just yet.

File: modules/javafx.graphics/src/test/java/test/javafx/stage/WindowTest.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/javafx/stage/WindowTest.java:222]: Oops, it is unintentional yeah.

Most of this patch was done with multi-caret editing to tackle multiple spots affected in the same way:
- Mark affected spot
- Ask my IDE to find and put a caret on all other occurrences of selected text
- Edit them all at once

Being in the rhythm of changing imports and `@Before` to `@BeforeEach` I probably unintentionally selected and changed this occurrence of the word "Before". I'll revert this.

File: modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceTestBase.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/javafx/concurrent/ServiceTestBase.java:65]: As mentioned in another comment, it's unlikely we'll spend more time on this later on. I will remove all the NOTE comments and commented out imports and such.

**Comment by lukostyra** [comment]
> should we set a minimum of two reviewers?

I agree, there should be another pair of eyes looking at this. Even though it's quite repetitive, it is quite a big change,

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java:76]: I'll leave it as-is then.

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/test/CssMethodsTestBase.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/test/CssMethodsTestBase.java:113]: Changed it to `Configuration` and all parameter method sources to `Stream<Configuration>`

File: modules/javafx.graphics/src/test/java/test/javafx/concurrent/ScheduledServiceTest.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/javafx/concurrent/ScheduledServiceTest.java:95]: Changed it to use `@BeforeEach` in non-parametrized classes

File: modules/javafx.graphics/src/test/java/test/javafx/scene/TreeShowingPropertyTest.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/TreeShowingPropertyTest.java:68]: You're right, someone probably did this originally for a reason. I didn't know `Supplier` would work like this, but the more you know!

Reverted to how it used to be, but JUnit5-ified.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/effect/BoxBlur_properties_Test.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/effect/BoxBlur_properties_Test.java:44]: Following above comments, I'll leave it as `Stream<Arguments>`

File: modules/javafx.graphics/src/test/java/test/javafx/scene/transform/AffineOperationsTest.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/transform/AffineOperationsTest.java:2781]: It definitely would be on the podium, especially for the parameter expansion... 181 tests times 32 arguments (5792 tests total).

But the undisputed champion is `TransformOperationsTest` with 81 tests times 116 arguments (9396 tests) =)

File: modules/javafx.graphics/src/test/java/test/javafx/scene/transform/TransformOperationsTest.java
Comment (lukostyra) [reply modules/javafx.graphics/src/test/java/test/javafx/scene/transform/TransformOperationsTest.java:2944]: It is either adding a `setUp(...)` line at the start of every function, or editing these. Hard to say which path is really better, so I'd stay with this one.

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java
Comment (andy-goryachev-oracle) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/sg/prism/DirtyRegionTestBase.java:76]: that's fine.
List is easier to manipulate, but this is very minor issue.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/TreeShowingPropertyTest.java
Comment (andy-goryachev-oracle) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/TreeShowingPropertyTest.java:79]: minor: this method should probably be `private`

File: modules/javafx.graphics/src/test/java/test/javafx/scene/layout/BackgroundTest.java
Comment (Maran23) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/layout/BackgroundTest.java:789]: Indentation looks off?

File: modules/javafx.graphics/src/test/java/test/javafx/scene/layout/BackgroundTest.java
Comment (Maran23) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/layout/BackgroundTest.java:796]: Indentation looks off?

File: modules/javafx.graphics/src/test/java/test/javafx/scene/layout/BorderPaneTest.java
Comment (Maran23) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/layout/BorderPaneTest.java:949]: Indentation looks off?

**Review by Maran23 — COMMENTED** [review]
Found some things that may should be improved, but looks good otherwise!

**Review by andy-goryachev-oracle — APPROVED** [review]
Looks good, thank you for making changes!
May thanks to @Maran23 for lending your expertise and time to this review!

**Review by Maran23 — APPROVED** [review]
Looks good!

**Comment by lukostyra** [comment]
/integrate

Thanks for your help everyone! This sure was a monster, but we got through it.

**Pull request closed without merging**
